### PR TITLE
Pixel Phase I - CA - improving fake rejection at high pt

### PR DESCRIPTION
**Pull request by felicepantaleo**
In the Cellular Automaton, we now require that, if two doublets are aligned in the xy-plane, they also have to be aligned with the beamspot ± (region radius + tolerance).
This has the effect of reducing the fake rate for pt > 10GeV/c, for quadruplets and especially for triplets.

@VinInn @rovere @makortel

### CODE REVIEW

**Comment by VinInn** [comment]
@cmsbuild , please test

**Comment by felicepantaleo** [comment]
You will find the effect of the PR in here:
https://fpantale.web.cern.ch/fpantale/offline_tuning/pull_request/index.html

harvest_original refers to the tracking which is in the release (no CA).
default_CA is the CA without the PR (with some tuned parameters for each iteration)
PR16491.root

**Comment by slava77** [comment]
sorry for the delay.
I'm going to check this today

**Comment by felicepantaleo** [comment]
@slava77, no problem. The MVA tuning is being done with this PR.

File: RecoPixelVertexing/PixelTriplets/plugins/CACell.h [hunk 1]
```diff
@@ -168,7 +168,14 @@ class CACell {
         float tan_12_13_half_mul_distance_13_squared = fabs(y1 * (x2 - x3) + y2 * (x3 - x1) + y3 * (x1 - x2)) ;
         if(tan_12_13_half_mul_distance_13_squared * ptmin <= 1.0e-4f*distance_13_squared)
         {
-        	return true;
+        	
+            float distance_3_beamspot_squared = (x3-region_origin_x) * (x3-region_origin_x) + (y3-region_origin_y) * (y3-region_origin_y);
+         
+            float dot_13_bs3_squared = (x1 - x3)*(x3 - region_origin_x) + (y1 - y3) * (y3 - region_origin_y);
+            float distance_13_beamspot_squared  = distance_3_beamspot_squared -  dot_13_bs3_squared;
```
Comment (slava77) [inline RecoPixelVertexing/PixelTriplets/plugins/CACell.h]: given that "1" is inner and "3" is outer, if I understand correctly, dot_13_bs3_squared is mostly negative.
So, I just wanted to double-check that the signs are correct (either in dot_13_bs3_squared or in this line where it's subtracted from distance_3_beamspot_squared).

Comment (felicepantaleo) [inline RecoPixelVertexing/PixelTriplets/plugins/CACell.h]: @slava77 
You are correct, this is incorrect. I'll correct, run validation  and come back with a fix.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@felicepantaleo 
please confirm that you are done with changes.
Thanks.

File: RecoPixelVertexing/PixelTriplets/plugins/CACell.h (hunk 1 above)
Comment (felicepantaleo) [inline RecoPixelVertexing/PixelTriplets/plugins/CACell.h]: @slava77 I was calling the variable "squared" but I was not squaring anything. 
Now we have a better fake rejection
https://fpantale.web.cern.ch/fpantale/offline_tuning/pull_request/plots_seeding_seeds_detachedTripletStep/dupandfake1.pdf

**Comment by felicepantaleo** [comment]
@slava77 
I confirm

File: RecoPixelVertexing/PixelTriplets/plugins/CACell.h (hunk 1 above)
Comment (slava77) [inline RecoPixelVertexing/PixelTriplets/plugins/CACell.h]: the pdf has less than 5% reduction in fake rate for pt>10 GeV (upper left).
Is this the improvement you are after?
[seems rather marginal]

Comment (felicepantaleo) [inline RecoPixelVertexing/PixelTriplets/plugins/CACell.h]: @slava77 some distributions of seeds:
https://fpantale.web.cern.ch/fpantale/offline_tuning/pull_request_highPtTriplet/ttbar_pu_building_highPtTripletStep/dist1.pdf

and 

https://fpantale.web.cern.ch/fpantale/offline_tuning/pull_request_detachedTriplet/ttbar_pu_seeding_seeds_detachedTripletStep/dist1.pdf

with no regression in timing and efficiency.
Feel free not to approve it if you think it's marginal. 
I would have wasted only 30 mins of work. ;-)

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/16491 c9758c1
- code looks OK. Reco is minimally affected at this point (CA is not used yet, so far only tuned up for HLT in private settings). Relying on MTV plots posted earlier by @felicepantaleo .
- jenkins tests pass

if HLT developments continue in 81X, this should be merged in 81X.

**Comment by davidlange6** [comment]
+1

**Pull request merged**